package/MDAnalysis/analysis/MSD.py

```diff
@@ -0,0 +1,398 @@
+from collections import deque
+
+import numpy as np
+import pickle
+
+import MDAnalysis
+from MDAnalysis.lib.distances import squared_distance_vector
+
+
+class MSD(object):
+    """Object for calculating mean square displacement (MSD)
+
+    Attributes
+    ----------
+    pos : list
+        contains position information, organized:
+        selection X frame X atom X (x, y, z)
+    atomids_per_sel_at_t : list
+        contains set of atoms which satisfy each selection at each
+        frame, organized:
+        selection X frame X set of atom ids
+    map_atomids_to_pos_at_t : list
+        contains dictionaries which map atom ids to corresponding
+        pos index, organized:
+        selection X frame X dict
+    msd : np.array(dtype=float64)
+        Contains the sum of squared displacements data, organized:
+        selection X tau
+        (MSD is calculated as msd/n_samples)
+    n_samples : np.array(dtype=int)
+        Contains the number of samples corresponding to each msd
+        entry, organized:
+        selection X tau
+
+    Methods
+    -------
+    _select(self)
+        Generates list of atom groups that satisfy select_list strings
+    _init_pos_deque(self):
+        Initializes lists necessary for MSD calculations at the first restart.
+    _update_deques(self, this_restart):
+        Updates lists necessary for MSD calculations
+    _process_pos_data(self):
+        Runs MSD calculation
+    run(self):
+        Calls _process_pos_data and saves final output
+
+    """
+
+    def __init__(self, universe, select_list, start_frame, stop_frame,
+                 dt_restart, out_name='msd', len_msd=250, write_output=True,
+                 max_data=False):
+        """
+        Parameters
+        ----------
+        universe : object
+            Universe object containing trajectory data
+        select_list : list
+            List of selection strings
+        start_frame : int
+            Time when analysis begins (in frames)
+        stop_frame : int
+            Time when analysis ends (in frames)
+        dt_restart : int
+            Spacing of restarts (in frames)
+        out_name: str
+            Name of pickle output file: (out_name).p
+        len_msd : int
+            Specifies the maximum tau to be calculated (frames)
+            (saves memory in array allocation)
+        write_output : bool
+            Specifies whether to write pickle file
+        max_data : bool
+            If true, include tau values after stop_frame in calculation (this
+            is useful for avoiding data loss when breaking large calculations
+            across multiple batch jobs)
+        """
+
+        self.universe = universe
+        self.select_list = select_list
+        if isinstance(select_list, str):
+            self.select_list = select_list.split(',')
+        print(self.select_list)
+        self.start_frame = int(start_frame)
+        self.stop_frame = int(stop_frame)
+        self.dt_restart = int(dt_restart)
```

Comment:
This will also mean that you need to introduce a `step` argument. Your `dt_restart` isn't really a step. But this is important to downsample a trajectory. Eg you sampled originally at 1ps but want to run the analysis for frames saved every 5 ps.